### PR TITLE
CodeFrames and coloring for Flow Parse Errors

### PR DESCRIPTION
**Pull request by samouri**
Addresses the first point of https://github.com/fastpack/fastpack/issues/95.

**summary**
Currently when FastPack encounters an error in a javascript source file, it outputs an error like this:

```
Project Directory: /.../test/error-parse
Mode: development
Call Stack:
        './a' from module: index.js
'./index.js' from module: $fp$main
Processing Module: a.js

Parse Error
File: a.js
        (1:15) - (1:16): Unexpected token ;
```

Ideally FastPack would output a codeframe that highlights exactly where in the code the issue is -- similar to how Elm or BuckleScript output errors.

### CODE REVIEW

**Comment by samouri** [comment]
I've taken some ideas from BuckleScript's output and currently have output like this:
screen shot 2018-07-18 at 5 57 07 pm

@zindel: do you have any ideas for what you'd like the error messages to look like?

File: Fastpack/Error.ml [hunk 1]
```diff
@@ -1,11 +1,47 @@
...
+  let endLine = min (List.length lines) (loc._end.line + 1) in
+  let codeframe = (List.filter_map (fun (i,line) -> 
+      if startLine <= i  && i <= endLine then 
+        Some (i, line) 
+      else None
+    ) lines) 
+  in
+  let formatted = List.map (fun (i, line) -> (
+        if loc.start.line <= i && i <= loc._end.line then 
+          let error_substring = String.sub line (loc.start.column) (loc._end.column - loc.start.column) in
```
Comment (zindel) [inline Fastpack/Error.ml]: You may want to use the `FastpackUtil.UTF8.sub` here

**Comment by zindel** [comment]
@samouri this looks great!

I don't have an exact idea how it should look like. I really like what I see on the attached screenshot! One thing I would probably add is a location in bold, so that it is a little bit more prominent and decoupled of the error message itself?

Couple related ideas to consider:

1. It would be great to make sure it works in both cases: when `stderr` is a tty or not. Maybe in a latter case we would want to use some ASCII art instead of colors? That'd also be useful together with the `--report=json` option. This way consumers won't have to deal with tty-specific color characters. For instance, the error for the example above may look like this:
    ```
    1 | // missing second half of addition
    2 | function hello() {
    3 |  a + ;
            ^
    4 | }
    ```
2. Please have the test with lots of Unicode characters in the source code. The problem is that `Loc.t` operates with Unicode symbols and not bytes. Hence `String.*` functions may not work as you expect. Take a look to `FastpackUtil/UTF8.ml` to get an idea.

**Comment by andreypopp** [comment]
@samouri that looks lovely!

**Comment by samouri** [comment]
> It would be great to make sure it works in both cases: when stderr is a tty or not

That makes sense to me, I'm on it!

> Please have the test with lots of Unicode characters in the source code.

Ah, I had no idea that the String.* functions don't work on Unicode.  I'll switch over to UTF8 functions like you suggest 👍

**Comment by samouri** [comment]
I made a couple updates so that it :

- handles unicode
- displays the filepath bolded
- prints caret and plaintext without ansi color codes when outputting places that are not a tty
- gracefully handles when the number of digits in the line number of the codeframe changes

File: Fastpack/GraphBuilder.ml [hunk 2]
```diff
@@ -183,23 +183,26 @@ let read_module
...
-             Lwt.fail (Context.PackError (ctx, CannotParseFile (location_str, args)))
-           | Preprocessor.Error message ->
-             Lwt.fail (Context.PackError (ctx, PreprocessorError message))
-           | FastpackUtil.Error.UnhandledCondition message ->
-             Lwt.fail (Context.PackError (ctx, UnhandledCondition message))
-           | exn ->
-             Lwt.fail exn
+            | FlowParser.Parse_error.Error args ->
+              let location_str = Module.location_to_string location in
+              let src = match source with
```
Comment (samouri) [inline Fastpack/GraphBuilder.ml]: In what situation do we not have the `source` for a file?

**Comment by samouri** [comment]
Have I been correct so far in using `./test/test.sh PATTERN` in order to run tests and `./test/update.sh PATTERN` in order to update the snapshots?

The one thing I haven't been able to figure out so far is how to write an automated test for the TTY case because the `.sh` scripts redirect stderr and stdout.

Besides for that wrinkle, I'd call this PR ready for review!
Its my first time trying write real OCaml code so feel free to be harsh in your feedback and also to push directly to this branch if you'd like.

Thanks!

File: Fastpack/Error.ml (hunk 1 above)
Comment (samouri) [reply Fastpack/Error.ml]: done

File: Fastpack/GraphBuilder.ml (hunk 2 above)
Comment (zindel) [reply Fastpack/GraphBuilder.ml]: The only case when source is None is when you have an `import/require` statement like this:
```javascript
import './some-loader!'
```
Quite rare case imho, but still possible.

File: Fastpack/Error.ml [hunk 3]
```diff
@@ -1,11 +1,58 @@
...
+      if startLine <= i  && i <= endLine then 
+        Some (i, line) 
+      else None
+    ) lines) 
+  in
+  let maxLineNo = List.fold_left (fun n (i, _) -> max n i) 0 lines in
+  let maxDigits = String.length (string_of_int maxLineNo) in
+  let formatted = List.map (fun (i, line) -> 
+      let isErrorLine = loc.start.line <= i && i <= loc._end.line in
+      let isTTY = (Unix.isatty Unix.stderr) in
```
Comment (zindel) [inline Fastpack/Error.ml]: Parenthesis are not needed here. I.e. `let isTTY = Unix.isatty Unix.stderr in` should work just fine

File: Fastpack/Error.ml [hunk 4]
```diff
@@ -1,11 +1,58 @@
...
+      let lineNo = String.pad (maxDigits) (string_of_int i) in
```
Comment (zindel) [inline Fastpack/Error.ml]: Same as above, just `maxDigits`

File: Fastpack/Error.ml [hunk 5]
```diff
@@ -1,11 +1,58 @@
...
+      match (isErrorLine, isTTY) with
```
Comment (zindel) [inline Fastpack/Error.ml]: Same as above: `match isErrorLine, isTTY with`

File: Fastpack/Error.ml [hunk 6]
```diff
@@ -1,11 +1,58 @@
...
+      | false, _ -> lineNo ^ " │ " ^ line 
+      | true, false -> 
+        let (offset, length) = (loc.start.column + 1), (loc._end.column - loc.start.column) in
```
Comment (zindel) [inline Fastpack/Error.ml]: Same as above: `let offset, length = (loc.start.column + 1), (loc._end.column - loc.start.column)`. They are not needed in the right-hand side as well, but I would probably keep them just for readability.

File: Fastpack/Error.ml [hunk 7]
```diff
@@ -1,11 +1,58 @@
...
+  let f = match font with 
+    | Regular -> "0"
+    | Bold -> "1"
+  in
+  "\027[" ^ f ^ ";" ^ col ^ "m" ^ str ^ "\027[0m"
+
+let get_codeframe (loc: Loc.t) lines = 
+  let startLine = max 0 (loc.start.line - 2) in
+  let endLine = min (List.length lines) (loc._end.line + 1) in
+  let codeframe = (List.filter_map (fun (i, line) ->
```
Comment (zindel) [inline Fastpack/Error.ml]: Outer-most parenthesis are obsolete, i.e. `let codeframe = List.filter_map ... lines in` would work fine

File: Fastpack/Error.ml [hunk 8]
```diff
@@ -36,20 +83,34 @@ let to_string package_dir error =
...
+  | CannotParseFile (location_str, errors, source) ->
+    let lines = String.split_on_char '\n' source
+                |> List.mapi (fun i line -> (i + 1, line)) in
+    let format_error (loc, error) = 
+      let error_desc = FlowParser.Parse_error.PP.error error in 
+      "--------------------\n"
+      ^ error_desc ^ " at " ^ (loc_to_string loc) ^ "\n"
+      ^ "\n"
+      ^ (get_codeframe loc lines)
+      ^ "\n"
```
Comment (zindel) [inline Fastpack/Error.ml]: You can use the `String.concat` in cases like this one: 
```OCaml
String.concat "\n" [
  "--------------------";
  error_desc ^ " at " ^ loc_to_string loc;
  get_codeframe loc lines;
  ""
]
```

**Comment by zindel** [comment]
@samouri, first, let me acknowledge the fact, that this is done very well and we'll happily merge it after the set of minor issues are addressed!

1. See the syntactic change requests above, those are really minor and not essential, but the code is a little bit more cleaner/easier to modify afterwards.

2. Regarding the `tty`-specific test. I think we can do the following:
    - let's have some secret environment variable, say `FPACK_TTY_TEST`
    - then we can wrap the `Unix.isatty` function which also respects the fact that this variable is set in the environment
    - the function should go to `FastpackUtil.FS.isatty. I am using `FastpackUtil.FS` for keeping all the operating system specific functions, so it should really be `OS` :) We can do the renaming later
    - finally in your `*.sh` test file you use it as `FPACK_TTY_TEST=yes $FPACK ...`. This will produce a little ugly snapshot, but in general will permit us to get the diff in future.

3. Somewhat related to the previous one. It looks like you're checking `isatty` in two places: error handling & inside of the `get_codeframe`. I would suggest to do it outside of the `get_codeframe` only and pass it in as an argument. In this case the function becomes completely "pure" and easier to deal with in other places.

Finally, the testing workflow question. Side note: I definitely have to properly document it :) Here is the anticipated workflow example:

```bash
$ make test-integration # run before committing
error-cannot-leave-package-dir/dev.sh [OK]
-------------------------------------------
error-cannot-rename-module-binding/dev.sh [OK]
-------------------------------------------
...
Total: 24. Failed: 0. Skipped: 24

$ make test-integration pattern=cra # handy, when developing some specific case
Packed in 0.967s. Bundle: 739Kb. Modules: 33.  Cache: empty. Mode: development.
pack-cra/dev.sh [OK]
-------------------------------------------
Total: 1. Failed: 0. Skipped: 24 Pattern mode. Consider running all the tests.

$ make train-integration pattern=cra # save snapshot, when you're satisfied with it
pack-cra/dev.sh [OK] Snapshot saved
-------------------------------------------
Total: 1. Failed: 0. Skipped: 24 Pattern mode. Consider running all the tests.

```

Thank you so much!

**Comment by zindel** [comment]
@andreypopp Well, as we've discussed before we're going to convert it to Reason later on, I didn't like `ocamlformat` when I tried it a few months ago. It didn't preserve empty lines at all and made the code quite unreadable

File: Fastpack/Error.ml (hunk 3 above)
Comment (samouri) [reply Fastpack/Error.ml]: done

File: Fastpack/Error.ml (hunk 4 above)
Comment (samouri) [reply Fastpack/Error.ml]: done

File: Fastpack/Error.ml (hunk 5 above)
Comment (samouri) [reply Fastpack/Error.ml]: done

File: Fastpack/Error.ml (hunk 6 above)
Comment (samouri) [reply Fastpack/Error.ml]: done

File: Fastpack/Error.ml (hunk 7 above)
Comment (samouri) [reply Fastpack/Error.ml]: done

File: Fastpack/Error.ml (hunk 8 above)
Comment (samouri) [reply Fastpack/Error.ml]: I applied this trick in a couple places.  The empty strings look a bit funny, but definitely still a readability boost :)

**Comment by samouri** [comment]
@zindel: I picked the name `FPACK_FORCE_TTY` for the environment variable since I like the ring of it a bit better.  Feel free to let me know if you still prefer `FPACK_TTY_TEST` so that its more obviously _just_ for the test case.

While the output of the test looks pretty strange if you simply `cat` it, I found a good trick for viewing it with colors: 
```bash
echo -ne $(cat ./test/error-parse-tty/dev/stderr.txt | sed  's/$/\\n/' | sed 's/ /
\\a /g')
```

Also, one other thing I found helpful is reviewing this pr [without whitespace changes](https://github.com/fastpack/fastpack/pull/96/files?w=1)

File: FastpackUtil/FS.ml [hunk 9]
```diff
@@ -60,3 +60,10 @@ let is_text_file filename =
   match Re.exec_opt pat_text_ext filename with
   | Some _ -> true
   | None -> false
+
+let isatty channel = 
+  let forceTTY = match Sys.getenv_opt "FPACK_FORCE_TTY" with
+    |  Some "true" -> true
```
Comment (zindel) [inline FastpackUtil/FS.ml]: Maybe `Some _` would work better? One may set the FPACK_FORCE_TTY=1

File: FastpackUtil/FS.ml [hunk 10]
```diff
@@ -60,3 +60,10 @@ let is_text_file filename =
   match Re.exec_opt pat_text_ext filename with
   | Some _ -> true
   | None -> false
+
+let isatty channel =
```
Comment (zindel) [inline FastpackUtil/FS.ml]: Overall it'd probably be more concise this way:
```OCaml
let isatty channel =
  match Sys.getenv_opt "FPACK_FORCE_TTY" with
  | Some _ -> true
  | None -> Unix.isatty channel
```

**Comment by zindel** [comment]
@samouri, thanks for making those updates! I agree with the `FPACK_FORCE_TTY ` naming. It definitely sounds more descriptive. I have 2 cosmetic comments for the `FS.isatty`, but those aren't important enough to hold the PR. So I'm happy to merge! Thank you & congratulations for the first PR merged! 🎉

**Pull request merged**